### PR TITLE
[nexus] Make the VPC creation saga unwind-safe, add tests

### PR DESCRIPTION
**Pull request by smklein**
- Defines undo actions for VPC creation
- Adds test for the VPC creation saga

Part of #2052

### CODE REVIEW

File: nexus/src/app/sagas/vpc_create.rs [hunk 1]
```diff
@@ -315,6 +374,21 @@ async fn svc_update_firewall(
...
+    sagactx: NexusActionContext,
+) -> Result<(), anyhow::Error> {
+    let osagactx = sagactx.user_data();
+    let params = sagactx.saga_params::<Params>()?;
+    let opctx = OpContext::for_saga_action(&sagactx, &params.serialized_authn);
+    let (authz_vpc, _) =
+        sagactx.lookup::<(authz::Vpc, db::model::Vpc)>("vpc")?;
+    osagactx
+        .datastore()
+        .vpc_update_firewall_rules(&opctx, &authz_vpc, vec![])
```
Comment (plotnick) [inline nexus/src/app/sagas/vpc_create.rs:387]: With the caveat that I don't really understand Sagas yet, this empty vector seems odd to me. Wouldn't we want an undo to restore the old set of rules, or at least the default rules? I don't think we default to an empty set anywhere else.

File: nexus/src/app/sagas/vpc_create.rs [hunk 2]
```diff
@@ -124,11 +129,6 @@ async fn svc_create_vpc(
     let vpc_id = sagactx.lookup::<Uuid>("vpc_id")?;
     let system_router_id = sagactx.lookup::<Uuid>("system_router_id")?;
 
-    // TODO: This is both fake and utter nonsense. It should be eventually
```
Comment (plotnick) [inline nexus/src/app/sagas/vpc_create.rs:127]: I am very glad this TODO can go away.

**Review by plotnick — APPROVED** [review]
Thank you for taking this on! It's great that we have the tooling in place to make these flows more robust. Modulo my concern about empty firewall rules in the undo action, this looks great to me.

File: nexus/src/app/sagas/vpc_create.rs (hunk 1 above)
Comment (smklein) [reply nexus/src/app/sagas/vpc_create.rs:387]: I believe the implementation of `vpc_update_firewall_rules` does the following:

- Deletes all firewall rules attached to the supplied `vpc_id`
- Inserts new rules, which reference `vpc_id`

This is why we're passing the empty vec - it clears out the old rules, and doesn't insert anything new.

Since this is an "undo" step along the pathway of VPC creation, I don't *think* it matters that we chuck the firewall rules -- we're making this call to delete all resources that might reference the VPC before we remove it.

TL;DR: I cleared it out to clean our SQL tables before removing the associated VPC. Does that seem okay?

File: nexus/src/app/sagas/vpc_create.rs [hunk 3]
```diff
@@ -333,3 +407,323 @@ async fn svc_notify_sleds(
...
+                datastore.pool_for_tests().await.unwrap(),
+            )
+            .await
+            .optional()
+            .unwrap()
+            .map(|rule| {
+                eprintln!("Firewall rule exists: {rule:?}");
+            })
+            .is_none()
+    }
```
Comment (smklein) [inline nexus/src/app/sagas/vpc_create.rs:649]: FYI, if I *don't* remove the firewall rules, this fails -- unwinding the saga would still leave old firewall rules around.

File: nexus/src/app/sagas/vpc_create.rs (hunk 1 above)
Comment (plotnick) [reply nexus/src/app/sagas/vpc_create.rs:387]: Thank you for that very helpful explanation; I see now. Yes, that seems fine.
